_P164_MSswitch.ino

```diff
@@ -0,0 +1,169 @@
+        addFormSelector(F("State"), F("state"), 2, sOptions, sOptionsValues, 0);
+        addFormCheckBox(F("Send on submit"), F("send"), false);
+      }
+      success = true;
+      break;
+    }
+
+    case PLUGIN_WEBFORM_SAVE:
+    {
+      PCONFIG_LONG(0) = getFormItemInt(F("p164_address"));
+      SocketAddress = PCONFIG_LONG(0);
+      if (isFormItemChecked(F("send"))) {
+        noInterrupts();
+        log164 = F("Sending: ");
+        noInterrupts();
+        if (getFormItemInt(F("state"))) {
+          if (getFormItemInt(F("channel"))==16) {
+            Plugin_164_msswitch->groupOn(SocketAddress);
+            log164 += F("GroupOn");
+          } else {
+            Plugin_164_msswitch->channelOn(SocketAddress,getFormItemInt(F("channel")));
+            log164 += F("channelOn: ");
+            log164 += getFormItemInt(F("channel"));
+            log164 += F(" Address: ");
+            log164 += SocketAddress;
+            log164 += F(" PIN: ");
+            log164 += CONFIG_PIN1;
+          }
+        } else {
+          if (getFormItemInt(F("channel"))==16) {
+            Plugin_164_msswitch->groupOff(SocketAddress);
+            log164 += F("GroupOff");
+          } else {
+            Plugin_164_msswitch->channelOff(SocketAddress,getFormItemInt(F("channel")));
+            log164 += F("channelOff: ");
+            log164 += getFormItemInt(F("channel"));
+          }
+        }
+        interrupts();
+        addLog(LOG_LEVEL_DEBUG, log164);
+      }
+      success = true;
+      break;
+    }
+
+    case PLUGIN_INIT:
+    {
+      if (!Plugin_164_msswitch)
+      {
+        Plugin_164_msswitch = new Socketswitch(CONFIG_PIN1);
+        log164 = F("Init 433MHZ Pin: ");
+        log164 += CONFIG_PIN1;
+        addLog(LOG_LEVEL_DEBUG, log164);
+      }
+      //SocketAddress = PCONFIG(0);
+      success = true;
+      break;
+    }
+
+    case PLUGIN_WRITE:
+    {
+      if (Plugin_164_msswitch)
+      {
+
+        SocketAddress = PCONFIG_LONG(0);
+        String tmpString  = string;
+        if (tmpString.startsWith(F("MSswitch"))){
+          noInterrupts();
+          int idx1 = tmpString.indexOf(',');
+          int idx2 = tmpString.indexOf(',', idx1+1);
+          int idx3 = tmpString.indexOf(',', idx2+1);
+          String val_dev = tmpString.substring(idx1+1, idx2);
+          String val_state = tmpString.substring(idx2+1, idx3);
+          if (val_dev.equalsIgnoreCase("g")) {
+            if (val_state.equalsIgnoreCase("on")) Plugin_164_msswitch->groupOn(SocketAddress);
+            if (val_state.equalsIgnoreCase("off")) Plugin_164_msswitch->groupOff(SocketAddress);
+          } else {
+            if (val_dev == "1" || val_dev == "2" || val_dev == "3" || val_dev == "4" || val_dev == "5" || val_dev == "6" || val_dev == "7" || val_dev == "8" || val_dev == "9" || val_dev == "10" || val_dev == "11" || val_dev == "12" || val_dev == "13" || val_dev == "14" || val_dev == "15" || val_dev == "16") {
```

Comment:
We have functions to check if a string is a numerical value. Please use that, then convert to an int and check its range.
These strings all add up to the memory consumption and binary file size.